bench/fixture.ts

```diff
@@ -0,0 +1,325 @@
+    // NOTE: The settlement contract is funded with a lot of extra tokens, so
+    // the settlement doesn't have to be perfect or balance out.
+
+    debug(`running fixture with ${JSON.stringify(options)}`);
+
+    const {
+      deployment: { settlement },
+      domainSeparator,
+      solver,
+      tokens,
+      traders,
+      uniswapPair,
+      uniswapTokens,
+    } = this;
+
+    const encoder = new SettlementEncoder(domainSeparator);
+
+    await tokens.ensureTokenCount(options.tokens);
+    for (let i = 0; i < options.trades; i++) {
+      // NOTE: Alternate the order flags, signing scheme and fee discount in
+      // such a way that the benchmark includes all possible combination of
+      // `(orderKind, partiallyFillable, signingScheme, feeDiscount)`.
+
+      let orderSpice: Pick<
+        Order,
+        "kind" | "partiallyFillable" | "sellAmount" | "buyAmount"
+      >;
+      switch (i % 4) {
+        case 0:
+          orderSpice = {
+            kind: OrderKind.SELL,
+            partiallyFillable: false,
+            sellAmount: ethers.utils.parseEther("100.0"),
+            buyAmount: ethers.utils.parseEther("99.0"),
+          };
+          break;
+        case 1:
+          orderSpice = {
+            kind: OrderKind.BUY,
+            partiallyFillable: false,
+            sellAmount: ethers.utils.parseEther("101.0"),
+            buyAmount: ethers.utils.parseEther("100.0"),
+          };
+          break;
+        case 2:
+          orderSpice = {
+            kind: OrderKind.SELL,
+            partiallyFillable: true,
+            sellAmount: ethers.utils.parseEther("200.0"),
+            buyAmount: ethers.utils.parseEther("199.0"),
+          };
+          break;
+        case 3:
+          orderSpice = {
+            kind: OrderKind.BUY,
+            partiallyFillable: true,
+            sellAmount: ethers.utils.parseEther("201.0"),
+            buyAmount: ethers.utils.parseEther("200.0"),
+          };
+          break;
+        default:
+          throw new Error("unreacheable");
+      }
+      const signingScheme =
+        (i + Math.floor(i / 4)) % 2 == 0
+          ? SigningScheme.TYPED_DATA
+          : SigningScheme.MESSAGE;
+      const feeDiscount = (i % 3) * (FULL_FEE_DISCOUNT / 2); // 0% | 50% | 100%
+
+      const dbg = {
+        fill: orderSpice.partiallyFillable
+          ? "partially fillable"
+          : "fill-or-kill",
+        kind: orderSpice.kind == OrderKind.SELL ? "sell" : "buy",
+        sign:
+          signingScheme == SigningScheme.TYPED_DATA ? "typed-data" : "message",
+        fee: 100 * (1 - feeDiscount / FULL_FEE_DISCOUNT),
+      };
+      debug(
+        `encoding ${dbg.fill} ${dbg.kind} order with ${dbg.sign} signature and ${dbg.fee}% fees`,
+      );
+
+      await encoder.signEncodeTrade(
+        {
+          sellToken: tokens.id(i % options.tokens).address,
+          buyToken: tokens.id(i % options.tokens).address,
```

Comment:
It's weird that we allow that the sell token is the same as the buy token. Should we require that this is not the case?

Reply:
This was a typo with my fixture! Nice catch.

As for
> Should we require that this is not the case?

I don't its needed. Maybe I'm missing something, but it would just be a user giving us free fees :smile:

Reply:
In principle it's right, as far as benching is concerned!
I proposed to require it in case a user signs an order with price < 1, which is a way for an evil solver to take free money (with the same token is used twice in the solution). I'm also ok with assuming that no user is going to manually sign such an order and save the cost of 1 require/order.

Reply:
> which is a way for an evil solver to take free money (with the same token is used twice in the solution).

Sure, but the user agreed to giving the solver free money :stuck_out_tongue:

Reply:
This is also causing unexpected gas results...